token/program-2022/src/extension.rs

```diff
@@ -1,6 +1,185 @@
 //! Extensions available to token mints and accounts
 
-use solana_program::{clock::Epoch, program_option::COption, pubkey::Pubkey};
+use {
+    crate::state::{pack_coption_key, unpack_coption_key, Account, Mint, Multisig},
+    arrayref::{array_mut_ref, array_ref},
+    solana_program::{
+        clock::Epoch,
+        program_error::ProgramError,
+        program_option::COption,
+        program_pack::{IsInitialized, Pack, Sealed},
+        pubkey::Pubkey,
+    },
+    std::convert::{TryFrom, TryInto},
+};
+
+const TYPE_LEN: usize = 1;
+const LENGTH_LEN: usize = 4;
+
+// TODO probably need an immutable version of this for clients
+/// Encapsulates base state data (mint or account) with possible extensions
+#[derive(Debug, PartialEq)]
+pub struct BaseStateWithExtensions<'data, S: BaseState> {
+    /// Unpacked base data
+    pub base: S,
+    /// Raw base data
+    pub base_data: &'data mut [u8],
+    /// Slice of data containing all TLV data, deserialized on demand
+    pub tlv_data: &'data mut [u8],
+}
+impl<'data, S: BaseState + Pack + IsInitialized> BaseStateWithExtensions<'data, S> {
```

Comment:
I really liked using Pods in zk token: https://github.com/solana-labs/spl-zk-token/blob/master/program/src/pod.rs

It's so much nicer than packing/unpacking: no copy, writes are immediate (no explicit pack needed). I'm game to make the requirement that all extensions be Pods.

Reply:
Yes please, I was hoping to start playing with bytemuck on this. Thanks for pointing out the helper types that you're using, we may even want to have those available in the SDK.  Would that make C integration really gross, or do we not really care?

Reply:
> we may even want to have those available in the SDK

Yeah for sure. I've already snuck a couple minimal Pod things into the sdk (e.g. https://github.com/solana-labs/solana/pull/20493). I've been waiting for more adoption before pulling in additional stuff, like these zk-token helpers, since it's hard to take stuff out of the sdk once added